### DISCUSSION TITLE
Feature suggestion: Improving Open WebUI "Plugin" Architecture to Detect Missing Dependencies/Breaking Change

### DISCUSSION THREAD

**Discussion by madslyng** [open-webui/open-webui#8180] [Ideas]
I’d like to raise a point about importing Functions that fail due to missing dependencies/breaking changes caused by changes to Open WebUI between versions; (see [Issue #8109](https://github.com/open-webui/open-webui/issues/8109)). We’ve all seen how confusing it can be to figure out whether the issue lies with Open WebUI itself or with a plugin/“function” when a dependency is missing, or for whatever reason is broken. This often leads to back-and-forth between users, plugin developers, and Open WebUI maintainers, which feels counterintuitive to the user-friendliness that Open WebUI aims for.

Suggestion: Let’s explore improved “plugin architecture” that handles version compatibility and breaking changes more gracefully. A system like this could alert users about incompatible or missing dependencies and guide them on updating plugins (or Open WebUI itself) to avoid those import failures. Such an approach could save time and make it simpler for everyone—users, plugin devs, and Open WebUI contributors—to address these issues.

Would love to hear your thoughts on whether this is feasible, and any ideas on how we might start building it.

**Comment by tjbck** [comment]
https://docs.openwebui.com/features/plugin/migration/

**Reply by madslyng** [reply]
@tjbck thanks for the link. Certainly a useful link, but doesn't address the problem from the user perspective. The user is left to try and figure out what is wrong. 

I'm thinking it probably wouldn't be a huge task to scan the code in the Function/Plugin, and detect if it is using incompatible dependencies.

That would be a massive UX improvement from a user perspective.

**Reply by tjbck** [reply]
Thanks for sharing your thoughts—it’s always great to see constructive ideas like this being brought forward. Before diving into specifics, I’d like to emphasize that Open WebUI is still very much in its exploratory phase, as it is currently in version 0.5.x. This means things are bound to evolve and change, especially in a field as rapidly advancing as the AI space. To some extent, this comes with the territory of working on a project that is breaking new ground.

That said, I completely agree with the intent behind your suggestion, and I share your focus on maintaining the user-friendliness that Open WebUI aims to deliver. Import failures due to missing dependencies or breaking changes in plugins can be very frustrating, especially for end-users who are simply trying to get things to work. While such issues aren’t necessarily everyday occurrences, and they will likely become less common as the project matures, even rare instances impact the user experience—and we should always be striving to make things better.

The good news is we are already thinking along these lines! A more robust plugin architecture that includes proactive measures—such as version compatibility checks, clearer error messages, or guidance for resolving dependencies—would indeed be a massive step forward. It’s feasible to imagine implementing something like a “compatibility scanner” that could validate function/plugin imports against the currently installed Open WebUI version and flag problems before they occur.

While building such a system would require careful planning and some additional development effort, it’s likely to pay dividends in terms of saving time—not only for users but also for plugin developers and maintainers who currently spend energy troubleshooting these back-and-forth issues. I see this as an excellent area where we could make a targeted improvement while still keeping our focus on iterating toward a stable 1.0.

For anyone reading this, we’d absolutely welcome PRs from contributors who are interested in helping make this a reality. If you have ideas for how to implement such a system—or even a partially complete approach—we’d love to see your input. Open WebUI thrives on contributions from the community, and suggestions like this serve as a great catalyst for innovation.

Thanks again for bringing this up—ideas like yours are what help make Open WebUI better. Let’s keep this conversation going. I’m curious to hear if you’ve thought of specific features this envisioned system should include, or if you’d like to collaborate on drafting a roadmap for this improvement.

**Reply by madslyng** [reply]
Happy to hear this is already being thought about - is there an issue that tracks this, so I can follow along? or where are the discussions and archtecture talks taking place ? I could naturally take a stab at some naive approaches as PRs as suggested. But i'd be very happy to follow the dialogue and chip in where possible
